### PR TITLE
adds some "factory" machines

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request

https://github.com/user-attachments/assets/4ceb4c0f-d5ef-4fc0-8436-d7eec5b6f396

https://github.com/user-attachments/assets/56ddd387-7376-4c35-a067-1adccbddeecd

https://github.com/user-attachments/assets/dda6cc2b-614a-4adb-a8f4-2c03b51162e0

https://github.com/user-attachments/assets/fa7697fb-f484-48a0-bb85-ee0c2f4867e2

https://github.com/user-attachments/assets/02de4b24-2fa0-4a1e-b147-df9500109b3c

https://github.com/user-attachments/assets/b56c03ab-49c9-487f-a99f-fcba5ce038ac

https://github.com/user-attachments/assets/52bae5a4-68b0-4f25-99c1-1b677b99790a

i didnt feel like recording the lathe and crafter for a suitable file size again but essentially the crafter crafts and the lathe lathes

all machines but the router and sorter are cable powered (suitable on lavaland)
theyre researched roundstart

they can receive any resource that bumps into it if that resource is on the conveyor

## Why It's Good For The Game
more fun engineering stuff
and perhaps mining given these are more efficient but require effort to set up
https://hackmd.io/@jimmyl/S1dZRZosC
## Changelog
:cl:
add: added the manufacturing smelter,router,sorter,crafter,lathe,crusher,unloader
/:cl:

### CODE REVIEW

**Comment by Whale-Splas** [comment]
If conveyor belt it self cannot work on field without apc,Why bother powering these machines with cables?

**Review by Wallemations — APPROVED** [review]
Oh my god I've been wanting stuff like this forever

Makes me wish we had the filter for disposal pipes too

**Comment by optimumtact** [comment]
no design doc or coherent overview of where the mechanic is intended to go.

**Comment by deathrobotpunch** [comment]
so what can you even make with this shit? circles and triangles?

**Comment by mc-oofert** [comment]
> no design doc or coherent overview of where the mechanic is intended to go or how it fits into the intended departments or existing game loop systems.

it is something you do for fun as it is 100% optional
anyone with engineering access may build a silly factory wherever they please

**Comment by Dawnseer** [comment]
Can I make it sort mail by department, or even first letter in recipients name?

**Comment by mc-oofert** [comment]
> Can I make it sort mail by department, or even first letter in recipients name?

there is a filter that checks if the items name contains the text you inputted
and one that also checks if its mail/package tagged for a specific room

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
So, we got a few action items.

This PR will **need** a design document. Some of the features included in this PR are cool and could be atomized out without any real issue:
* Sorter
* Router
* Possibly Unloader (I'm going to mull it over, it may or may not be fine)

Some of them currently intersect with existing design spaces:
* Crusher 

It being a combined arcmining machine as well as a recycler without any reduction in efficiency strikes me as a possible issue, as in theory it's a faster round start smelter/refinery from arcmining without mining point collection. I'm not blocking this, but I would need to know your design intent on it first.

Economy & Resource Concerns (I know I'm sorry):
*The Manu-Crafter and Manu-Lathe

These could in theory be used to mass produce large quantities of materially cheap resources that have static export values like tools or wooden crates. Now, for the former after some basic discussion, that makes for a better justification for removing tool export values than it does than blocking an automatic lathe machine, but it begs a bigger question of "is there anything in the lathe that you actually need as such quantities that you would need to automate it"? And I'd need to hear that from that design doc

Also, please include a full description of all the aspects of your PR in the PR body knowwhatimsaying

(Your PR kicks ass but we're gonna do it right)

**Comment by deathrobotpunch** [comment]
this is gonna need a guide to factory in the wiki

**Comment by RikuTheKiller** [comment]
> ArcaneMusic

Atomizing this PR sounds like an unnecessary pain since the features are built to work with one another and use the same backend, so they'd be dependent on a factory machine backend PR either way.

I agree with you that the Crusher should be unable to process boulders. Jimmy outlined that it's because his machines can be powered via wires but I don't know why you couldn't just add that capability to the arcmining refinery. Because you can and in this case, should.

As for the Manu-Lathe and Manu-Crafter, the economy allowing you to sell static exports like that is the issue. Especially with shit like coffins. You should NOT be able to export them for money. Crate recycling is meant to be, well, recycling, not manufacturing.

As for what you'd manufacture with a crafter, anything from upgraded weapons to spears to IEDs is on the table. I.e. it's entirely possible to set up a crafter line to go from glass and iron into spears. And I think that's pretty damn cool. If you think there aren't enough craftable items that you can use en masse, then just add more. Either way, those are the items I could come up with without looking at the crafting menu at all. I figure there's others.

As for the lathe, it can be used to make components for the crafter. Kind of wish it would allow departmental encryption keys or something so you could actually make advanced weapon manufacturies with no external inputs if security permitted it or something, but oh well.

**Comment by deathrobotpunch** [comment]
1984

**Comment by ArcaneMusic** [comment]
Posting here so I can provide a review after work 👍

**Comment by Ghommie** [comment]
> Posting here so I can provide a review after work 👍

Yet you forgot to...

File: code/modules/manufactorio/machines/sorter.dm [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+/obj/machinery/power/manufacturing/sorter/Initialize(mapload)
+	. = ..()
+	if(isnull(dir_if_not_met))
+		dir_if_not_met = dir
+	var/static/list/loc_connections = list(
+		COMSIG_ATOM_ENTERED = PROC_REF(on_entered),
+	)
+	AddElement(/datum/element/connect_loc, loc_connections)
+	for(var/i=1, i <= length(sort_filters), i++)
```
Comment (Ghommie) [inline code/modules/manufactorio/machines/sorter.dm:26]:
```suggestion
	for(var/i in 1 to length(sort_filters))
```
The operations inside the loop don't seem to affect the length of sort_filters or do anything fancy on the loop itself, so there's no need to explicify the do statement I guess.

File: code/modules/manufactorio/machines/unloader.dm [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		if(!as_closet.can_open())
+			return MANUFACTURING_FAIL
+	else if(!istype(as_orebox))
+		return MANUFACTURING_FAIL
+	receiving.Move(src, get_dir(receiving, src))
+	START_PROCESSING(SSfastprocess, src)
+	return MANUFACTURING_SUCCESS
+
+/obj/machinery/power/manufacturing/unloader/process(seconds_per_tick)
+	if(!length(contents - circuit))
```
Comment (Ghommie) [inline code/modules/manufactorio/machines/unloader.dm:47]: yeah, define a list var.

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/modules/manufactorio/machines/unloader.dm (hunk 2 above)
Comment (mc-oofert) [reply code/modules/manufactorio/machines/unloader.dm:47]: A list var cant be done without modifying it again later everywhere; most machines cause contents to be modified during processing and usually check it after that again

File: code/game/turfs/turf.dm [hunk 3]
```diff
@@ -270,7 +270,7 @@ GLOBAL_LIST_EMPTY(station_turfs)
  * * type_list - are we checking for types of atoms to ignore and not physical atoms
  */
 /turf/proc/is_blocked_turf(exclude_mobs = FALSE, source_atom = null, list/ignore_atoms, type_list = FALSE)
-	if(density)
+	if((!isnull(source_atom) && !CanPass(source_atom, get_dir(src, source_atom))) || density)
```
Comment (LemonInTheDark) [inline code/game/turfs/turf.dm:273]: what the fuck?

Comment (mc-oofert) [reply code/game/turfs/turf.dm:273]: forgor

**Pull request merged**